### ISSUE TITLE
Where a navigation property is specified in $select, the corresponding navigation link is not represented in the response

### ISSUE THREAD

**Issue by gathogojr** [OData/AspNetCoreOData#944]
**Assemblies affected**
ASP.NET Core OData 8.2.0

**Describe the bug**
Where a navigation property is specified in `$select`, the corresponding navigation link is not represented in the response.
https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionselect

**Reproduce steps**
Build a simple OData service as follows:
```csharp
// Data model
public class Customer
{
    public int Id { get; set; }
    public List<Order>? Orders { get; set; }
}

public class Order
{
    public int Id { get; set; }
}

// Controller
public class CustomersController
{
    private static readonly List<Customer> customers = new List<Customer>(
        Enumerable.Range(1, 2).Select(idx => new Customer
        {
            Id = idx,
            Orders = new List<Order>(
                Enumerable.Range(1, 2).Select(dx => new Order
                {
                    Id = dx
                }))
        }));

    [EnableQuery]
    public ActionResult<IEnumerable<Customer>> Get()
    {
        return customers;
    }
}

// Edm model and service configuration
using Microsoft.AspNetCore.OData;
using Microsoft.OData.ModelBuilder;

var builder = WebApplication.CreateBuilder(args);

var modelBuilder = new ODataConventionModelBuilder();
modelBuilder.EntitySet<Customer>("Customers");
modelBuilder.EntitySet<Order>("Orders");

builder.Services
    .AddControllers()
    .AddOData(
        options => options.EnableQueryFeatures().AddRouteComponents(
            model: modelBuilder.GetEdmModel()));

var app = builder.Build();

app.UseRouting();
app.UseEndpoints(endpoints => endpoints.MapControllers());

app.Run();
```

**Data Model**
Included in _Reproduce steps_

**EDM (CSDL) Model**
```xml
<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <Schema Namespace="NS.Models" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityType Name="Customer">
                <Key>
                    <PropertyRef Name="Id" />
                </Key>
                <Property Name="Id" Type="Edm.Int32" Nullable="false" />
                <NavigationProperty Name="Orders" Type="Collection(NS.Models.Order)" />
            </EntityType>
            <EntityType Name="Order">
                <Key>
                    <PropertyRef Name="Id" />
                </Key>
                <Property Name="Id" Type="Edm.Int32" Nullable="false" />
            </EntityType>
        </Schema>
        <Schema Namespace="Default" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityContainer Name="Container">
                <EntitySet Name="Customers" EntityType="NS.Models.Customer">
                    <NavigationPropertyBinding Path="Orders" Target="Orders" />
                </EntitySet>
                <EntitySet Name="Orders" EntityType="NS.Models.Order" />
            </EntityContainer>
        </Schema>
    </edmx:DataServices>
</edmx:Edmx>
```

**Request/Response**
Request:
```http
http://localhost:5237/Customers?$select=Orders
```
Response:
```json
{
    "@odata.context": "http://localhost:5237/$metadata#Customers(Orders)",
    "value": [
        {},
        {}
    ]
}
```

**Expected behavior**
The empty braces should contain the corresponding navigation links
```json
{
    "@odata.context": "http://localhost:5237/$metadata#Customers(Orders)",
    "value": [
        {"Orders":"http://localhost:5237/Customers/1/Orders"},
        {"Orders":"http://localhost:5237/Customers/2/Orders"}
    ]
}
```

**Additional context**
Bug is reproduceable is [7.x](https://github.com/OData/WebApi) - Port fix.

**Comment by xuzhg** [comment]
@gathogojr What's the value of **SelectedNavigationProperties** on **SelectExpandNode** when doing serialization?

and What's the value of https://github.com/OData/AspNetCoreOData/blob/ea4e79ce3a16238f12f55ca34c23a46cc01e291c/src/Microsoft.AspNetCore.OData/Formatter/Serialization/ODataResourceSerializer.cs#L1166
and it will be called within 'WriteNavigationLinksAsync'.

**Comment by gathogojr** [comment]
Thank you @xuzhg for the hint. I followed the logic for writing the navigation link. I found out that the `BuildNavigationLink` method (referenced by `CreateNavigationLink`) returns `null` because one condition is not fulfilled for the navigation link to be generated:
https://github.com/OData/AspNetCoreOData/blob/e7211260f2bb2bcb76b3b1de744d0c42565a6131/src/Microsoft.AspNetCore.OData/Edm/NavigationSourceLinkBuilderAnnotation.cs#L214-L216

The condition `!navigationLinkBuilder.FollowsConventions` evaluates to `false` and for that reason we don't execute https://github.com/OData/AspNetCoreOData/blob/e7211260f2bb2bcb76b3b1de744d0c42565a6131/src/Microsoft.AspNetCore.OData/Edm/NavigationSourceLinkBuilderAnnotation.cs#L218

The comment
https://github.com/OData/AspNetCoreOData/blob/e7211260f2bb2bcb76b3b1de744d0c42565a6131/src/Microsoft.AspNetCore.OData/Edm/NavigationSourceLinkBuilderAnnotation.cs#L221 suggests that ODL should save the day if a `null` is returned but ODL does no such thing. ODL ignores writing the navigation link if the value is `null`.

I get the expected output if I remove the condition `!navigationLinkBuilder.FollowsConventions` but then, `CreateNavigationLink` method is referenced in multiple places.
Why do we restrict use of navigation link builder situations to where it **doesn't** follow OData conventions?

**Comment by gathogojr** [comment]
The _problem_ starts in ODL.

When we call the `InnerPrepareResourceForWriteStart` method, the `CreateResourceMetadataBuilder` method on the following line will return a null for `JsonMininalMetadataLevel`:
https://github.com/OData/odata.net/blob/3133a34c7c5d2006e04059655cb258fb23c81309/src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs#L2404

Because of the reason stated here: https://github.com/OData/odata.net/blob/3133a34c7c5d2006e04059655cb258fb23c81309/src/Microsoft.OData.Core/JsonLight/JsonMinimalMetadataLevel.cs#L56-L59

When the above happens, we'll have the `MetadataBuilder` property of the `ODataResource` initialized to `NoOpResourceMetadataBuilder` that returns `null` when `GetNavigationLinkUrl` method is called.

If the `MetadataBuilder` property was set to `ODataConventionalEntityMetadataBuilder`, deferring to ODL to generate the navigation link works.

In summary, the issue stems from the fact that we're writing what is basically an odata annotation against `odata.metadata=mininal` and `JsonMinimalMetadataLevel` doesn't create the metadata builder that is capable of generating the expected navigation link
